### PR TITLE
Add vfs diff, update tsbuild test

### PR DESCRIPTION
**Pull request by rbuckton**
With the recent change to project references, @weswigham and I noticed a few issues with sourcemap emit. To better diagnose these, I've added the ability to generate a diff between two virtual filesystems so that we can get a better idea of what changes we are making. This change will help diagnose source map paths in #24917.

### CODE REVIEW

File: src/testRunner/unittests/tsbuild.ts [hunk 1]
```diff
@@ -394,32 +395,27 @@ namespace ts {
...
-
-    function loadFsMirror(vfs: vfs.FileSystem, localRoot: string, virtualRoot: string) {
-        vfs.mkdirpSync(virtualRoot);
-        for (const path of Harness.IO.readDirectory(localRoot)) {
-            const file = getBaseFileName(path);
-            vfs.writeFileSync(virtualRoot + "/" + file, Harness.IO.readFile(localRoot + "/" + file)!);
-        }
-        for (const dir of Harness.IO.getDirectories(localRoot)) {
-            loadFsMirror(vfs, localRoot + "/" + dir, virtualRoot + "/" + dir);
+        function patchResolver(resolver: vfs.FileSystemResolver): vfs.FileSystemResolver {
```
Comment (weswigham) [inline src/testRunner/unittests/tsbuild.ts]: Is this absolutely necessary? The mounted files are lazily loaded now, right? And only diffs are in the output? What still depends on only these lib files being findable?

File: package.json [hunk 2]
```diff
@@ -45,7 +45,7 @@
         "@types/minimatch": "latest",
         "@types/minimist": "latest",
         "@types/mkdirp": "latest",
-        "@types/mocha": "latest",
+        "@types/mocha": "^5.2.2",
```
Comment (weswigham) [inline package.json]: Should probably still be `latest`.

File: src/harness/vfs.ts [hunk 3]
```diff
@@ -671,6 +671,160 @@ namespace vfs {
...
+            // no difference if both buffers are identical
+            if (Buffer.compare(changedBuffer, baseBuffer) === 0) return false;
+
+            container[basename] = new File(changedBuffer);
+            return true;
+        }
+
+        private static symlinkDiff(container: FileSet, basename: string, changedNode: SymlinkInode, baseNode: SymlinkInode) {
+            // no difference if the nodes are the same reference
+            if (changedNode.symlink === baseNode.symlink) return false;
```
Comment (weswigham) [inline src/harness/vfs.ts]: If symlinks are interpreted as relative paths, should a symlink at `/lib/foo` of `../bar` be equivalent to one at the same location of `../../lib/bar`? Or am I mistaken?

File: src/harness/vfs.ts [hunk 4]
```diff
@@ -1201,6 +1350,24 @@ namespace vfs {
...
+    export class Rmdir {
+        private _rmdirBrand?: never; // brand necessary for proper type guards
+    }
+
+    /** Unlinks a file in a `FileSet` */
+    export class Unlink {
+        private _unlinkBrand?: never; // brand necessary for proper type guards
+    }
+
+    // prevents compiler errors due to unused brand properties
```
Comment (weswigham) [inline src/harness/vfs.ts]: Don't we have some `_unlinkBrand?!` or `["_unlinkBrand"]` syntax variant that should more cleanly suppress that error? Or is that just definite assignment?

File: src/harness/vfs.ts (hunk 3 above)
Comment (rbuckton) [reply src/harness/vfs.ts]: This compares symlinks in the same path, so you shouldn't run into this.

**Pull request merged**